AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/HierarchicalVersionUtils.dfy

```diff
@@ -215,4 +214,27 @@ module {:options "/functionSyntax:4" } HierarchicalVersionUtils {
     else
       Failure("Unable to decode string")
   }
+
+  function HierarchyVersionToString(version: Types.HierarchyVersion): (output: string)
+    ensures Structure.StringIsValidHierarchyVersion?(output)
+    ensures  StringToHierarchyVersion(output) == version
+  {
+    match version {
+      case v1 => Structure.HIERARCHY_VERSION_VALUE_1
+      case v2 => Structure.HIERARCHY_VERSION_VALUE_2
+    }
+  }
+
+  function StringToHierarchyVersion(version: string): (output: Types.HierarchyVersion)
+    requires Structure.StringIsValidHierarchyVersion?(version)
+    ensures version == Structure.HIERARCHY_VERSION_VALUE_1 ==> output.v1?
+    ensures version == Structure.HIERARCHY_VERSION_VALUE_2 ==> output.v2?
+  {
+    if version == Structure.HIERARCHY_VERSION_VALUE_1 then Types.HierarchyVersion.v1 else Types.HierarchyVersion.v2
```

Comment:
Could/should this be safer? What happens if version is neither v1 nor v2?
I understand this passes proof due to the nature of StringIsValidHierarchyVersion, so not blocking

Reply:
Oh, so the risk is only if a user (us or an external user) manages to call the Dafny transpiled logic of this method directly, as Dafny will never allow Dafny logic to call this method without meeting the requires clause.

Personally, I am fine with this risk, as we have already accepted it everywhere we use a `requires` clause. 

Using Dafny without using `requires` is... well... it would be difficult.
I did not want to throw an error here,
as refactoring `Structure.dfy` and `MutationStateStructures.dfy` to account for an error result would be terribly expensive.

Reply:
I'm more thinking about the risk of us adding HV3, 4, 5... and forgetting to update this logic

Reply:
The Ensures clause handle that.
Line 231 can be read out loud as:
"The version string is equal to the string '2', therefore the hierarchy-version is 2".

If we added an HV3, line 231 would fail `dafny verify`.